application-numbered-content-headings-api/src/main/java/org/xwiki/contrib/numbered/content/headings/internal/HeadingNumberingCalculator.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.contrib.numbered.content.headings.internal;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to calculate the number of a heading based on previously added headings.
+ *
+ * @version $Id$
+ * @since 1.2
+ */
+public class HeadingNumberingCalculator
```

Comment:
Probably best to make this an internal component (with `@Component(roles = HeadingNumberingCalculator.class)`)

Reply:
I don't see why. Instances of this class should never be re-used and are also not thread-safe. As the numbering services are singletons, there therefore shouldn't be an instance variable of this type in `AbstractHeadingsNumberingService`. We would thus always need to go through the component manager explicitly to get a fresh instance of this class. Is there any benefit from doing this over just creating a new instance of this simple helper class manually?

Reply:
Indeed probably not useful since it's internal. In this case a comment explaining why it's structured that way might help.

Reply:
Also, even though the code looks good, could be interesting to have a corresponding test class.